sdk/state/payment_test.go

```diff
@@ -13,195 +14,97 @@ import (
-				},
-				ProposerSignatures: CloseSignatures{
-					Close: []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
-				},
-			},
-			false,
-		},
-		{
-			CloseEnvelope{
-				Details: CloseDetails{
-					ObservationPeriodTime:      time.Minute,
-					ObservationPeriodLedgerGap: 2,
-					IterationNumber:            3,
-					Balance:                    100,
-					ConfirmingSigner:           keypair.MustParseAddress("GCJFS4LZFAM7NKFQFEWE4W2SCGARSE2SMLGNWGHH2LSZ6CLX326MZWPO"),
-				},
-				ProposerSignatures: CloseSignatures{
-					Close: []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
-				},
-			},
-			CloseEnvelope{
-				Details: CloseDetails{
-					ObservationPeriodTime:      time.Minute,
-					ObservationPeriodLedgerGap: 2,
-					IterationNumber:            3,
-					Balance:                    100,
-					ConfirmingSigner:           nil,
-				},
-				ProposerSignatures: CloseSignatures{
-					Close: []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
-				},
-			},
-			false,
-		},
+func TestCloseDetails_Equal(t *testing.T) {
+	assert.True(t, CloseDetails{}.Equal(CloseDetails{}))
+
+	// The same value should be equal.
+	ft := time.Now().UnixNano()
+	od1 := CloseDetails{}
+	fuzz.NewWithSeed(ft).NilChance(0).Fuzz(&od1)
+	t.Log("od1:", od1)
+	od2 := CloseDetails{}
+	fuzz.NewWithSeed(ft).NilChance(0).Fuzz(&od2)
+	t.Log("od2:", od2)
+	assert.True(t, od1.Equal(od2))
+
+	// Different values should never be equal.
+	for i := 0; i < 20; i++ {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			f := fuzz.New()
+			a := CloseSignatures{}
+			f.Fuzz(&a)
+			t.Log("a:", a)
+			b := CloseSignatures{}
+			f.Fuzz(&b)
+			t.Log("b:", b)
+			assert.False(t, a.Equal(b))
+			assert.False(t, b.Equal(a))
+		})
+	}
+}
+
+func TestCloseSignatures_Equal(t *testing.T) {
+	assert.True(t, CloseSignatures{}.Equal(CloseSignatures{}))
+
+	// The same value should be equal. It is common for CloseSignatures to be
+	// defined in whole or not at all, so we test that use case.
+	ft := time.Now().UnixNano()
+	os1 := CloseSignatures{}
+	fuzz.NewWithSeed(ft).NilChance(0).Fuzz(&os1)
+	t.Log("os1:", os1)
+	os2 := CloseSignatures{}
+	fuzz.NewWithSeed(ft).NilChance(0).Fuzz(&os2)
+	t.Log("os2:", os2)
+	assert.True(t, os1.Equal(os2))
+
+	// Different values should never be equal.
+	for i := 0; i < 20; i++ {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			f := fuzz.New()
+			a := CloseSignatures{}
+			f.Fuzz(&a)
+			t.Log("a:", a)
+			b := CloseSignatures{}
+			f.Fuzz(&b)
+			t.Log("b:", b)
+			assert.False(t, a.Equal(b))
+			assert.False(t, b.Equal(a))
+		})
 	}
-	for i, tc := range testCases {
+}
+
+func TestCloseEnvelope_Equal(t *testing.T) {
+	assert.True(t, CloseEnvelope{}.Equal(CloseEnvelope{}))
+
+	// The same value should be equal. It's common for CloseEnvelopes to start
+	// with details then have signatures added, so we check that pattern of
+	// incrementally adding fields.
+	f := fuzz.New().NilChance(0)
```

Comment:
💭  in some spots we're using `fuzz.New()` and in others `fuzz.NewWithSeed()`. Can we keep them the same for consistency?

according to [here](https://github.com/google/gofuzz/blob/f12f061189cca81b834eba9985016bc70e1dce2e/fuzz.go#L52) New() implicitly calls NewWithSeed with `time.Now().UnixNano()`, so maybe only using New()?

Reply:
The places I'm using `NewWithSeed()` I'm using a common seed between multiple fuzzers. This means the data the fuzzer uses is identical. This means I can construct two `OpenDetails` that are entirely independent objects, even internal pointers are different, however they should be the same in terms of equality and the Equal function.

The places I'm using `New()` I don't need any consistency between the fuzzers, or rather I want inconsistency. I could use `NewWithSeed()` in those places and manually pass in different times, but that seems unnecessary?

Reply:
ah that makes sense, thx for explaining. Agreed how it is lgtm